clinto/parsers/base.py

```diff
@@ -0,0 +1,63 @@
+from __future__ import absolute_import
+import json
+import os
+import copy
+
+
+def update_dict_copy(a, b):
+    temp = copy.deepcopy(a)
+    temp.update(b)
+    return temp
+
+
+class ClintoArgumentParserException(Exception):
+    def __init__(self, parser, *args, **kwargs):
+        self.parser = parser
+
+
+def parse_args_monkeypatch(self, *args, **kwargs):
+    raise ClintoArgumentParserException(self)
+
+
+class BaseParser(object):
+
+    extensions = []
+    contains = ""
+
+    def __init__(self, script_path=None, script_source=None):
+        self.is_valid = False
+        self.error = ''
+        self.parser = None
+
+        self.script_path = script_path
+        self.script_source = script_source
+
+        if not self.check_valid():
+            return
+
+        self.extract_parser()
+
+        if self.parser is None:
+            return
+
+        self.process_parser()
+
+    def check_valid(self):
+        if self.script_path:
+            ext = os.path.splitext(os.path.basename(self.script_path))[1]
+            return ext in self.extensions and self.contains in self.script_source
```

Comment:
But to test which parser you have you're going to have to get the parser first. That means for a click script you have to first try and process it as argparse - and that has to happen for every script. It's not so bad now but once we've implemented other command parsers (and languages) it's going to be a lot of wasted processing on those scripts.

I'm more looking for a way to prefer one parser over another (e.g. try and locate ClickParser first for scripts that *look like* click scripts). I'll put together an implementation of the heuristic thingy to demonstrate it .... for example it could be as simple as "does the file contain 'python' or 'import' or 'lambda' ...then it's more likely to be a Python file than R, test the Python loaders first".

I'm getting convinced that we can sidestep AST altogether by monkey patching and throwing exceptions - it works without problems on every argparse + click script I've tested. The advantage of doing it this way around is that once you have the object you *know* what it is, because otherwise you wouldn't have it. The only way you can get a result from monkey-patching argparse is if the parser is argparse! :)